File: src/domain/locations.ts
import type { GeoPoint } from "./types";

const KNOWN_ORIGINS: Array<{ keywords: string[]; point: GeoPoint }> = [
  { keywords: ["北京通州", "通州"], point: { lat: 39.9169, lng: 116.6462 } },
  { keywords: ["北京南站"], point: { lat: 39.865, lng: 116.379 } },
  { keywords: ["德州东站", "德州"], point: { lat: 37.443, lng: 116.374 } },
  { keywords: ["天津南站", "天津"], point: { lat: 39.0622, lng: 117.0669 } },
  { keywords: ["济南西站", "济南"], point: { lat: 36.6683, lng: 116.892 } },
  { keywords: ["上海虹桥", "虹桥"], point: { lat: 31.194, lng: 121.318 } }
];

export function resolveKnownOrigin(label: string): GeoPoint | undefined {
  const normalized = label.trim().toLowerCase();

  return KNOWN_ORIGINS.find((origin) =>
    origin.keywords.some((keyword) => normalized.includes(keyword.toLowerCase()))
  )?.point;
}

File: CarRentalOptimizer/Domain/Locations.swift
import Foundation

private struct KnownOrigin {
    let keywords: [String]
    let point: GeoPoint
}

private let knownOrigins: [KnownOrigin] = [
    KnownOrigin(keywords: ["北京通州", "通州"], point: GeoPoint(lat: 39.9169, lng: 116.6462)),
    KnownOrigin(keywords: ["北京南站"], point: GeoPoint(lat: 39.865, lng: 116.379)),
    KnownOrigin(keywords: ["德州东站", "德州"], point: GeoPoint(lat: 37.443, lng: 116.374)),
    KnownOrigin(keywords: ["天津南站", "天津"], point: GeoPoint(lat: 39.0622, lng: 117.0669)),
    KnownOrigin(keywords: ["济南西站", "济南"], point: GeoPoint(lat: 36.6683, lng: 116.892)),
    KnownOrigin(keywords: ["上海虹桥", "虹桥"], point: GeoPoint(lat: 31.194, lng: 121.318))
]

func resolveKnownOrigin(_ label: String) -> GeoPoint? {
    let normalized = label.trimmingCharacters(in: .whitespaces).lowercased()
    return knownOrigins.first { origin in
        origin.keywords.contains { keyword in
            normalized.contains(keyword.lowercased())
        }
    }?.point
}
